packages/components/affix/src/affix.vue

```diff
@@ -26,101 +31,92 @@ const ns = useNamespace('affix')
   }
 })
 
 const update = () => {
-  if (!root.value || !target.value || !scrollContainer.value) return
+  if (!scrollContainer.value) return
 
-  const rootRect = root.value.getBoundingClientRect()
-  const targetRect = target.value.getBoundingClientRect()
-  height.value = rootRect.height
-  width.value = rootRect.width
   scrollTop.value =
     scrollContainer.value instanceof Window
       ? document.documentElement.scrollTop
       : scrollContainer.value.scrollTop || 0
-  clientHeight.value = document.documentElement.clientHeight
 
   if (props.position === 'top') {
-    if (props.target) {
-      const difference = targetRect.bottom - props.offset - height.value
-      fixed.value = props.offset > rootRect.top && targetRect.bottom > 0
-      transform.value = difference < 0 ? difference : 0
-    } else {
-      fixed.value = props.offset > rootRect.top
-    }
-  } else {
     if (props.target) {
       const difference =
-        clientHeight.value - targetRect.top - props.offset - height.value
+        targetRect.bottom.value - props.offset - rootRect.height.value
       fixed.value =
-        clientHeight.value - props.offset < rootRect.bottom &&
-        clientHeight.value > targetRect.top
-      transform.value = difference < 0 ? -difference : 0
+        props.offset > rootRect.top.value && targetRect.bottom.value > 0
+      transform.value = difference < 0 ? difference : 0
     } else {
-      fixed.value = clientHeight.value - props.offset < rootRect.bottom
+      fixed.value = props.offset > rootRect.top.value
     }
+  } else if (props.target) {
+    const difference =
+      windowHeight.value -
+      targetRect.top.value -
+      props.offset -
+      rootRect.height.value
+    fixed.value =
+      windowHeight.value - props.offset < rootRect.bottom.value &&
+      windowHeight.value > targetRect.top.value
+    transform.value = difference < 0 ? -difference : 0
+  } else {
+    fixed.value = windowHeight.value - props.offset < rootRect.bottom.value
   }
 }
 
-const onScroll = () => {
-  update()
-
+const handleScroll = () => {
   emit('scroll', {
     scrollTop: scrollTop.value,
     fixed: fixed.value,
   })
 }
 
-watch(fixed, (fixed) => {
-  emit('change', fixed)
-})
+watch(fixed, () => emit('change', fixed.value))
 
 onMounted(() => {
   if (props.target) {
     target.value =
       document.querySelector<HTMLElement>(props.target) ?? undefined
-    if (!target.value) {
-      throw new Error(`Target is not existed: ${props.target}`)
-    }
+    if (!target.value)
+      throwError(COMPONENT_NAME, `Target is not existed: ${props.target}`)
   } else {
     target.value = document.documentElement
   }
   scrollContainer.value = getScrollContainer(root.value!, true)
+  rootRect.update()
 })
 
-useEventListener(scrollContainer, 'scroll', onScroll)
-useResizeObserver(root, () => update())
-useResizeObserver(target, () => update())
+useEventListener(scrollContainer, 'scroll', handleScroll)
+watchEffect(update)
```

Comment:
Do we have perf concern here? Since every reactive value is tracked by `watchEffect`

Reply:
I think it's acceptable...